### PR TITLE
[WIP] Feat/vtc

### PR DESCRIPTION
**Pull request by hadware**
This is a working PR on the future VTC implementation inspired from @MarvinLvn 's work, and to be merged into the next release of pyannote-audio.

Note: nothing has been done yet, this is just to get things started.

### CODE REVIEW

**Comment by hbredin** [comment]
FYI: https://github.com/pyannote/pyannote-audio/blob/fe29b430e0f3380a1471c6ba155f165aabf8096f/tutorials/add_your_own_task.ipynb

**Comment by hadware** [comment]
All right, It's currently working on my test dataset, I'll try to test it some more on my fake data, and then on some real data (@MarvinLvn  's and our clinical data) to see if it matches (or hopefully even beats) the former scores.

File: pyannote/audio/pipelines/multilabel_detection.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        self.segmentation_inference_ = Inference(model, **inference_kwargs)
+
+        self.binarize_hparams = ParamDict(**{
+            class_name: ParamDict(
+                onset=Uniform(0., 1.),
+                offset=Uniform(0., 1.),
+                min_duration_on=Uniform(0., 2.),
+                min_duration_off=Uniform(0., 2.),
+                pad_onset=Uniform(-1., 1.),
+                pad_offset=Uniform(-1., 1.)
```
Comment (hbredin) [inline pyannote/audio/pipelines/multilabel_detection.py:120]: In relation with https://github.com/pyannote/pyannote-pipeline/pull/34,  this is a good use case for freezing only parts of `ParamDict`. `pad_onset` and `pad_offset` are seldom useful and it makes sense to reduce the dimension of the hyperparameter search space by freezing them to 0.

Comment (hadware) [reply pyannote/audio/pipelines/multilabel_detection.py:120]: Yes! I wondered if `pad_{off,on}set` was relevant or not, since you weren't using it in the VAD pipeline. I'll freeze that once it's freezable. (What I could also do is not parameterize it though?)

**Comment by hadware** [comment]
I actually thought a bit about things, and It needs a bit more tweaking of the "custom" "MultilabelFScore" metric i've implemented for the optimization part.

**Comment by hbredin** [comment]
Hey @hadware, the `stale` bot seems unhappy with the status of this PR... Any update on your side?

**Comment by hadware** [comment]
Ah! Well, I'm not planning on giving up on this, but I'm currently a bit overwhelmed by some other work. The dust should settle by mid december, but I might need some input from you in the meantime to write a proper test of the new pipeline: how do you advise that I test it now that the `pyannote-audio` CLI endpoint is "deprecated"?

**Comment by hbredin** [comment]
> Ah! Well, I'm not planning on giving up on this, but I'm currently a bit overwhelmed by some other work. The dust should settle by mid december, but I might need some input from you in the meantime to write a proper test of the new pipeline: how do you advise that I test it now that the `pyannote-audio` CLI endpoint is "deprecated"?

Tasks are currently tested by simply checking that training is not broken (i.e. does not raise any Exception). 
This is done by running notebooks available in `/notebook/` directory. You could add your task to the `example.ipynb` notebook for example.

As far as pipelines are concerned, they are simply not tested for now 👎

**Comment by hadware** [comment]
Sorry for the "supremely concise" commit history, I tried to rebase on your develop branch but really bad things happened. Git is still an unforgiving ~master~ main for me, it seems.

**Comment by hadware** [comment]
Small question needed for model testing: what would you advise that I use for checkpointing? Pytorch-lightning's way? I can't really figure out how you're "expecting" it to be done in Pyannote's current iteration.

**Comment by hadware** [comment]
Nevermind, I figured I should use the `pyannote-audio-train` CLI endpoint.

**Comment by hadware** [comment]
Sorry for the dumbi-ish question, but i've never used Hydra on my own so I'm not very familiar with it. Here's the problem:
- from what i've understood, since i've added a `VoiceTypeClassification.yaml` "subconfig" to pyannote, i'm mostly good to go
- from what i've understood, if users want to "override" some of theses values, they can just do so using the CLI
- however, there are some parameters that is mandatory for this task and that _cannot_ have default values and _shouldn't be_ defined in the CLI: classes, classes unions and classes intersections
- in my opinion, these should be defined in a user-specified YAML file, that should be specified when running the pipeline via the CLI, but the YAML config should be "overloading" the one already specified for the `VoiceTypeClassification` task

So here's my question:
- what's the structure for such a yaml config file
- how do you feed it to hydra's CLI 
(sorry that's actually two questions, but you get the point).

**Comment by hbredin** [comment]
Hmmm. I am not sure why you need to use the CLI for testing the model.
This should be enough: 

```python
task = VoiceTypeClassification(...)
model = PyanNet(task=task)
trainer = Trainer(max_epochs=1)
trainer.fit(model)
```

Also, what do you mean by "model testing"?

**Comment by hadware** [comment]
Sorry, by testing i meant "using the new VTC task + pipeline on our data to see if it reproduces our well-known results, or even beats them". Thus this means training, validation and testing. I mostly need help on the training part (which I intend to do using the CLI, c.f. my previous post).

**Comment by hbredin** [comment]
Don't bother using the CLI -- try something like this:

```python
from pyannote.database import get_protocol
dataset = get_protocol('YourDataset.SpeakerDiarization.YourProtocol) 

from pyannote.audio.tasks import VoiceActivityDetection
vad = VoiceActivityDetection(dataset)

from pyannote.audio.models.segmentation import PyanNet
model = PyanNet(task=vad, sincnet={"stride": 10})

from pytorch_lightning.callbacks import EarlyStopping
from pytorch_lightning.callbacks.model_checkpoint import ModelCheckpoint
from pytorch_lightning.loggers import TensorBoardLogger

value_to_monitor, min_or_max = vad.val_monitor

model_checkpoint = ModelCheckpoint(
    monitor=value_to_monitor, 
    mode=min_or_max, 
    save_top_k=5, 
    every_n_epochs=1, 
    save_last=True, 
    dirpath=".", 
    filename=f"{{epoch}}-{{{value_to_monitor}:.6f}}",
    verbose=True)

early_stopping = EarlyStopping(
    monitor=value_to_monitor,
    mode=min_or_max,
    min_delta=0.0,
    patience=10.,
    strict=True,
    verbose=False) 

logger = TensorBoardLogger(".", name="", version="", log_graph=False)

from pytorch_lightning import Trainer
trainer = Trainer(gpus=1, callbacks=[model_checkpoint, early_stopping], logger=logger)
trainer.fit(model)
```

**Comment by hadware** [comment]
Thanks a lot, this was very helpful!

I have some good news: the whole pipeline seems to be working and giving out nice results. Here's a small table that i'm going to update as we run more tests and might fix some things:

Pyan. Vers. | Model | Dataset | Data Augment. | Best Epoch | Tuning Metric | Fscore | IER
-- | -- | -- | -- | -- | -- | -- | --
V1 | Pyannet | Clinical Itws | Noise (MUSAN) | 100 | Fscore | 86.6 | 19.6
V2 | Pyannet | Clinical Itws | None | 38 | IER | 86.80 | 20.69
V2 | Pyannet | Clinical Itws | None | TBA | Fscore | TBA | TBA


Note: on V1, I used the cyclic rate scheduler, as per @MarvinLvn 's worthy advice.
